src/dataset/mod.rs

```diff
@@ -134,3 +134,62 @@ mod tests {
         assert_eq!(*m[1][3], 9);
     }
 }
+
+impl<X: Copy + std::fmt::Debug, Y: Copy + std::fmt::Debug> std::fmt::Display for Dataset<X, Y> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        struct Target<Y> {
+            name: String,
+            value: Y,
+        }
+        struct Feature<X> {
+            name: String,
+            value: X,
+        }
+        struct DataPoint<X, Y> {
+            labels: Vec<Target<Y>>,
+            features: Vec<Feature<X>>,
+        }
+        impl<X: Copy + std::fmt::Debug, Y: Copy + std::fmt::Debug> std::fmt::Display for DataPoint<X, Y> {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                write!(
+                    f,
+                    "{} : {}",
+                    self.labels
+                        .iter()
+                        .map(|target| format!("{}:{:?}, ", target.name, target.value))
+                        .collect::<String>(),
+                    self.features
+                        .iter()
+                        .map(|feature| format!("{}:{:?}, ", feature.name, feature.value))
```

Comment:
I am wondering, how will the proposed message format fits large datasets? Let's say we have 2K samples in a dataset. Do we need to repeat feature name for every value we display? It seems redundant to me...
Another thought, should we truncate the message when certain number of samples has been reached?

Reply:
You could imagine a CSV-like format, But as someone who always chooses expanded output in Postgres, I'm not really a fan. It works really well until rows won't fit on a single line on the screen, and then it's awful. You could imagine that we intelligently choose depending on number of features vs. number of data points, which could be better, but not sure how it would work in practice. Maybe we should detect screen width vs num_features and decide based on that?

Truncating the message seems like a straightforward thing to do. Like, say, after 200 lines or something? I bet someone would want it configurable, though.

How do alternatives like Pandas (or whatever, I'm not familiar with data science tools) do it?

Reply:
I see your point here, Tormod. It is always frustrating when you have more attributes per sample than your screen can fit and data shifts by one line making it harder to skim through the data. But I don't think that printing feature name along with every value makes it easier to read the data either. I've played with your code a little and made it print Iris dataset. This is what I see, please correct me it this is not how you intended the format to be in this PR: 

```
Iris dataset: https://archive.ics.uci.edu/ml/datasets/iris
setosa:0.0, versicolor:0.0, virginica:0.0, :0.0,  : sepal length (cm):5.1, sepal width (cm):3.5, petal length (cm):1.4, petal width (cm):0.2, 
setosa:0.0, versicolor:0.0, virginica:0.0, :0.0,  : sepal length (cm):4.9, sepal width (cm):3.0, petal length (cm):1.4, petal width (cm):0.2, 
setosa:0.0, versicolor:0.0, virginica:0.0, :0.0,  : sepal length (cm):4.7, sepal width (cm):3.2, petal length (cm):1.3, petal width (cm):0.2, 
setosa:0.0, versicolor:0.0, virginica:0.0, :0.0,  : sepal length (cm):4.6, sepal width (cm):3.1, petal length (cm):1.5, petal width (cm):0.2, 
setosa:0.0, versicolor:0.0, virginica:0.0, :0.0,  : sepal length (cm):5.0, sepal width (cm):3.6, petal length (cm):1.4, petal width (cm):0.2, 
```

One problem with this way of displaying data is that it is hard to focus on numbers due to repeated feature names. Also, if we have more attributes per sample data will still take more lines per sample thus making it harder to see where one sample ends and another starts.

Pandas displays data as a table, that is truncated to `max_rows` and `max_columns`:

![image](https://user-images.githubusercontent.com/1102497/109203646-44808480-7759-11eb-95cd-f4fcbb451fda.png)